src/manifest.rs

```diff
@@ -33,15 +33,212 @@ use crate::{
 /// The name of the manifest file
 pub const MANIFEST_FILE: &str = "Proto.toml";
 
-/// A `buffrs` manifest format used for serialization and deserialization.
+/// The canary edition supported by this version of buffrs
+pub const CANARY_EDITION: &str = concat!("0.", env!("CARGO_PKG_VERSION_MINOR"));
+
+/// Edition of the buffrs manifest
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Edition {
+    /// The canary edition of manifests
+    ///
+    /// This indicates that breaking changes and unstable behavior can occur
+    /// at any time. Users are responsible for consulting documentation and
+    /// help channels if errors occur.
+    Canary,
+    /// Unknown edition of manifests
+    ///
+    /// This is unrecommended as breaking changes could be introduced due to being
+    /// in the beta release channel
+    Unknown,
+}
+
+impl Default for Edition {
+    fn default() -> Self {
+        Self::Canary
+    }
+}
+
+/// A buffrs manifest format used for serialization and deserialization.
 ///
 /// This contains the exact structure of the `Proto.toml` and skips
 /// empty fields.
-#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
-struct RawManifest {
-    package: Option<PackageManifest>,
-    #[serde(default)]
-    dependencies: DependencyMap,
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum RawManifest {
+    Canary {
+        package: Option<PackageManifest>,
+        dependencies: DependencyMap,
+    },
+    Unknown {
+        package: Option<PackageManifest>,
+        dependencies: DependencyMap,
+    },
+}
+
+impl RawManifest {
+    fn package(&self) -> Option<&PackageManifest> {
+        match self {
+            Self::Canary { package, .. } => package.as_ref(),
+            Self::Unknown { package, .. } => package.as_ref(),
+        }
+    }
+
+    fn dependencies(&self) -> &DependencyMap {
+        match self {
+            Self::Canary { dependencies, .. } => dependencies,
+            Self::Unknown { dependencies, .. } => dependencies,
+        }
+    }
```

Comment:
Now all of this logic needs to be duplicated, and will have to be changed again for `v1`, but I can't understand it's justification.

Reply:
It is not possible to have a:

```
struct RawManifest {
  edition: Option<Edition>,
  package: Option<..
  ..
}
```

here as the editions may contain different fields / data.